platforms/software/dependency-management/src/main/java/org/gradle/internal/component/AbstractVariantSelectionException.java

```diff
@@ -16,17 +16,43 @@
 
 package org.gradle.internal.component;
 
+import com.google.common.collect.ImmutableList;
+import org.gradle.internal.exceptions.ResolutionProvider;
 import org.gradle.internal.exceptions.StyledException;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+// TODO: Doc all the types in this hierarchy to note they should only be created by the ResolutionFailureHandler.  Perhaps make their constructors package-private?
 /**
  * Abstract base class for all attribute matching selection failures occurring at any stage of dependency resolution.
  */
-public abstract class AbstractVariantSelectionException extends StyledException {
+public abstract class AbstractVariantSelectionException extends StyledException implements ResolutionProvider {
+    private final List<String> resolutions = new ArrayList<>(1); // Usually there is only one resolution
+
     public AbstractVariantSelectionException(String message) {
-        super(message);
+        this(message, null);
     }
 
-    public AbstractVariantSelectionException(String message, Throwable cause) {
+    public AbstractVariantSelectionException(String message, @Nullable Throwable cause) {
         super(message, cause);
     }
+
+    /**
+     * Adds a resolution to the list of resolutions.
+     *
+     * Meant to be called during subclass construction, so <strong>must</strong> remain safe to do so by only accessing fields on this type,
+     * hence the {@code final} modifier.
+     *
+     * @param resolution The resolution (suggestion message) to add
+     */
+    public final void addResolution(String resolution) {
```

Comment:
IMO these exceptions should be immutable and these resolutions should be provided in the constructor

Reply:
I'm going to try and get this fully working first, then can make a pass, either in this PR or the next, to make them immutable.  I just don't want to keep churning on the structure of these exceptions - adding and removing doc registry was bad enough.